### PR TITLE
docs(lsp): more precise type annotations

### PR DESCRIPTION
**Pull request by tomtomjhj**

### CODE REVIEW

File: runtime/doc/lsp.txt [hunk 1]
```diff
@@ -697,9 +696,8 @@ buf_request_sync({bufnr}, {method}, {params}, {timeout_ms})
                       result. Defaults to 1000
 
     Return: ~
-        table<integer, any>|nil result, string|nil err Map of
-        client_id:request_result. On timeout, cancel or error, returns `(nil,
-        err)` where `err` is a string describing the failure reason.
+        table<integer, {err: lsp.ResponseError , result: any}>|nil result, string|nil err Map of
+        client_id:request_result. On timeout, cancel or error, returns `(nil, err)` where `err` is a string describing the failure reason.
```
Comment (mfussenegger) [inline runtime/doc/lsp.txt:700]: Looks like this causes failures with the vimdoc parser:

```
FAILED   test/functional/lua/help_spec.lua @ 13: :help docs validate
test/functional/lua/help_spec.lua:25: expected no parse errors, got: 1

stack traceback:
	test/functional/lua/help_spec.lua:25: in function <test/functional/lua/help_spec.lua:13>
```

It doesn't like the `|` 

Using `:InspectTree` on the help file shows the `ERROR` node:

image

File: runtime/lua/vim/lsp.lua [hunk 2]
```diff
@@ -2022,9 +2022,10 @@ end
 ---@param bufnr (integer) Buffer handle, or 0 for current.
 ---@param method (string) LSP method name
 ---@param params (table|nil) Parameters to send to the server
----@param callback (function) The callback to call when all requests are finished.
+---@param callback (fun(request_results: table<integer, {error: lsp.ResponseError, result: any}>))
```
Comment (folke) [inline runtime/lua/vim/lsp.lua:2025]: There is special syntax to make this work with vimdoc.

```
@param fun(request_results: table<integer, {error: lsp.ResponseError, result: any}>) (function) nlabla
```

Same with @return

You can always put `(vimdoctype)` in the comment, so after the type definition. That will then be used as the type in the vimdoc instead.

For more examples, check shared.lua

File: runtime/lua/vim/lsp.lua [hunk 3]
```diff
@@ -2067,7 +2068,8 @@ end
 ---@param timeout_ms (integer|nil) Maximum time in milliseconds to wait for a
 ---                               result. Defaults to 1000
 ---
----@return table<integer, any>|nil result, string|nil err Map of client_id:request_result.
+---@return table<integer, {err: lsp.ResponseError, result: any}>|nil result, string|nil err
```
Comment (folke) [inline runtime/lua/vim/lsp.lua:2071]: same here

File: runtime/lua/vim/lsp.lua [hunk 4]
```diff
@@ -1058,7 +1058,7 @@ function lsp.start_client(config)
   --- Returns the default handler if the user hasn't set a custom one.
   ---
   ---@param method (string) LSP method name
-  ---@return function|nil The handler for the given method, if defined, or the default from |vim.lsp.handlers|
+  ---@return lsp-handler|nil The handler for the given method, if defined, or the default from |vim.lsp.handlers|
```
Comment (folke) [inline runtime/lua/vim/lsp.lua:1061]: Where is this type defined?

**Review by folke — CHANGES_REQUESTED** [review]
Type annotations are not correct and some types seems to be missing.

If you change the types, you have to make sure that they work for:
- vimdoc
- lua_ls

File: runtime/lua/vim/lsp.lua (hunk 2 above)
Comment (tomtomjhj) [reply runtime/lua/vim/lsp.lua]: Thanks for the info. Fixed.

File: runtime/lua/vim/lsp.lua (hunk 3 above)
Comment (tomtomjhj) [reply runtime/lua/vim/lsp.lua]: Fixed

File: runtime/lua/vim/lsp.lua (hunk 4 above)
Comment (tomtomjhj) [reply runtime/lua/vim/lsp.lua:1040]: https://github.com/neovim/neovim/blob/236c20795eb9f11e21e0719b735ea741711acc08/runtime/lua/vim/lsp/types.lua#L3

File: runtime/doc/lsp.txt (hunk 1 above)
Comment (clason) [reply runtime/doc/lsp.txt]: Exactly. Try separating the `|` with spaces (which is not legal for taglinks, which this conflicts with).

If this becomes too cumbersome, we could switch to `foo?` for optional parameters.

Comment (tomtomjhj) [reply runtime/doc/lsp.txt]: Fixed as suggested by folke. The remaining test failures don't seem to be related to this.

**Pull request merged**